src/zenml/integrations/kubernetes/orchestrators/kubernetes_orchestrator_entrypoint.py

```diff
@@ -162,24 +172,96 @@ def run_step_on_kubernetes(step_name: str) -> None:
             mount_local_stores=mount_local_stores,
         )
 
-        # Create and run pod.
-        core_api.create_namespaced_pod(
-            namespace=args.kubernetes_namespace,
-            body=pod_manifest,
-        )
+        try:
+            # Create and run pod.
+            core_api.create_namespaced_pod(
+                namespace=args.kubernetes_namespace,
+                body=pod_manifest,
+            )
+        except Exception:
+            logger.error(
+                f"Failed to create pod for step `{step_name}`. Exiting."
+            )
+
+            raise
 
         # Wait for pod to finish.
         logger.info(f"Waiting for pod of step `{step_name}` to start...")
-        kube_utils.wait_pod(
-            kube_client_fn=lambda: orchestrator.get_kube_client(
-                incluster=True
-            ),
-            pod_name=pod_name,
-            namespace=args.kubernetes_namespace,
-            exit_condition_lambda=kube_utils.pod_is_done,
-            stream_logs=True,
+        try:
+            kube_utils.wait_pod(
+                kube_client_fn=lambda: orchestrator.get_kube_client(
+                    incluster=True
+                ),
+                pod_name=pod_name,
+                namespace=args.kubernetes_namespace,
+                exit_condition_lambda=kube_utils.pod_is_done,
+                stream_logs=True,
+            )
+
+            logger.info(f"Pod for step `{step_name}` completed.")
+        except Exception:
+            logger.error(f"Pod for step `{step_name}` failed.")
+
+            raise
+
+    def finalize_run(node_states: Dict[str, NodeStatus]) -> None:
+        """Finalize the run.
+
+        Args:
+            node_states: The states of the nodes.
+        """
+        # Some steps may have failed because the pods could not be created.
+        # We need to check for this and mark the step run as failed if so.
+
+        # Fetch the pipeline run using any means possible.
+        list_args: Dict[str, Any] = {}
+        if args.run_id:
+            # For a run triggered outside of a schedule, we can use the
+            # placeholder run ID to find the pipeline run.
+            list_args = dict(id=UUID(args.run_id))
+        else:
+            # For a run triggered by a schedule, we can only use the
+            # orchestrator run ID to find the pipeline run.
+            list_args = dict(orchestrator_run_id=args.run_name)
```

Comment:
I'm not sure this `args.run_name` is the correct value to use. As far as I can see, it might get sanitized before being used as the pod name later and will therefore not always be the `orchestrator_run_id`.

However, this entrypoint file is actually responsible for setting the orchestrator run ID, and we currently use `orchestrator_run_id = socket.gethostname()` to do so. So I think it's best if we use that exact value again here to make sure this always works.

Orchestrator run ID in combination with the deployment ID is actually also covered by a unique constraint in the DB, so in theory we don't even need the passing of the placeholder run as this will work also in non-scheduled cases.

Reply:
I tried to use `orchestrator_run_id` in one of my earlier attempts, but the `orchestrator_run_id` is not set for a placeholder run (unless at least one step has already run ??)

Reply:
I switched to using `orchestrator_run_id`